### PR TITLE
perf: Reduce time taken by metrics collection

### PR DESCRIPTION
**Pull request by underyx**
PR checklist:

- [ ] Documentation is up-to-date
- [ ] Changelog is up-to-date
- [ ] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

**Comment by underyx** [comment]
~gotta set up daemonized subprocess stuff here~

never mind, not possible

File: semgrep/semgrep/app/session.py [hunk 1]
```diff
@@ -31,13 +33,43 @@ class UserAgent:
...
+                ],
+                cwd=Path(__file__).parent,
+                stderr=subprocess.DEVNULL,
+                encoding="utf-8",
+            ).strip()
+            # If we installed semgrep in `.venv/` in the semgrep-docs repo,
+            # this function would report semgrep-docs commit SHAs in the user agent.
+            # This is why we verify the origin URL, and why check with .endswith()
+            if remote_url.replace(".git", "").endswith("/semgrep"):
+                result.add(f"sha/{sha}")
```
Comment (underyx) [inline semgrep/semgrep/app/session.py:64]: I thought this'd be handy to be able to recognize when weird metrics come from faulty commits, and to filter out metrics from development in general. But it's highly optional.

File: semgrep/semgrep/metrics.py [hunk 2]
```diff
@@ -280,14 +280,11 @@ def send(self) -> None:
         self._sent_at = datetime.now().astimezone().isoformat()
 
         try:
-            r = app_session.post(
+            r = requests.post(
```
Comment (underyx) [inline semgrep/semgrep/metrics.py:283]: We dropped the retry policy by no longer using the app session.

File: semgrep/semgrep/app/session.py [hunk 3]
```diff
@@ -31,13 +33,43 @@ class UserAgent:
...
+        except subprocess.CalledProcessError:
+            pass
+
+        return result
 
     def __str__(self) -> str:
         result = f"{self.name}/{self.version}"
-        for note in sorted(self.tags):
+        for note in sorted(self.tags, key=lambda x: ("/" in x, x)):
```
Comment (brendongo) [inline semgrep/semgrep/app/session.py:72]: is note supposed to be a Tuple(bool, str)

Comment (underyx) [reply semgrep/semgrep/app/session.py:72]: nope, this is just the sorting key; essentially what this does is:

1. first, order all notes to the front that have no `/` in them
2. then, order the rest alphabetically

Comment (underyx) [reply semgrep/semgrep/app/session.py:72]: so this leads to this ordering: `(Docker) (command/scan) (sha/123987)`

Comment (underyx) [reply semgrep/semgrep/app/session.py:72]: CleanShot 2022-05-24 at 15 39 20@2x

sample in practice

File: semgrep/semgrep/app/session.py (hunk 1 above)
Comment (brendongo) [reply semgrep/semgrep/app/session.py:64]: eh seems special case-y. Is it difficult to filter it out server side?

File: semgrep/semgrep/app/session.py (hunk 3 above)
Comment (brendongo) [reply semgrep/semgrep/app/session.py:72]: I see

File: semgrep/semgrep/app/session.py (hunk 1 above)
Comment (underyx) [reply semgrep/semgrep/app/session.py:64]: without this we don't have any markers that'd let us know if a metric was sent from one of our devs

**Review by brendongo — APPROVED** [review]
Would rather not have the semgrep url filtering but don't feel too strongly

File: semgrep/semgrep/app/session.py (hunk 1 above)
Comment (underyx) [reply semgrep/semgrep/app/session.py:64]: so it's impossible to filter; but we can also just decide not to care about filtering

**Pull request merged**